tests/components/switch/test_init.py

```diff
@@ -72,3 +74,183 @@ async def test_switch_context(
+
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_ON
+
+    # Turn the switch off when it's on
+    context = core.Context(user_id=hass_admin_user.id)
+    await hass.services.async_call(
+        "switch",
+        "turn_off",
+        {"entity_id": "switch.test1"},
+        True,
+        context,
+    )
+    # Make sure the switch is not optimistic
+    prev_state = state
+    state = hass.states.get("switch.test1")
+    assert state is prev_state
+
+    # Force update state to on, the state change should not be attributed to the context
+    prev_state = state
+    entity.async_set_state(STATE_ON)
+    state = hass.states.get("switch.test1")
+    assert state is not prev_state
+    assert state.state == STATE_ON
+    assert state.context is not context
+
+    # Change state to off, the state change should be attributed to the context
+    entity.async_set_state(STATE_OFF)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_OFF
+    assert state.context is context
+
+    # Change state back to on, the state change should not be attributed to the context
+    entity.async_set_state(STATE_ON)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_ON
+    assert state.context is not context
+
+    # Change state to off again, the state change should be attributed to the context
+    entity.async_set_state(STATE_OFF)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_OFF
+    assert state.context is context
+
+    # Bump time and change state, the state change should not be attributed to the context
+    freezer.tick(timedelta(seconds=5.001))
+    entity.async_set_state(STATE_OFF)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_OFF
+    assert state.context is not context
+
+    # Turn the switch on when it's off
+    context = core.Context(user_id=hass_admin_user.id)
+    await hass.services.async_call(
+        "switch",
+        "turn_on",
+        {"entity_id": "switch.test1"},
+        True,
+        context,
+    )
+
+    # Force update state to off, the state change should not be attributed to the context
+    prev_state = state
+    entity.async_set_state(STATE_OFF)
+    state = hass.states.get("switch.test1")
+    assert state is not prev_state
+    assert state.state == STATE_OFF
+    assert state.context is not context
+
+    # Change state to on, the state change should be attributed to the context
+    entity.async_set_state(STATE_ON)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_ON
+    assert state.context is context
+
+    # Change state back to off, the state change should not be attributed to the context
+    entity.async_set_state(STATE_OFF)
+    state = hass.states.get("switch.test1")
+    assert state.state == STATE_OFF
+    assert state.context is not context
+
+    # Change state to on again, the state change should be attributed to the context
```

Comment:
I'm trying to think of a use case where the same call to switch.turn_on causes your device to turn on twice in 5s.

It seems like once you get a match the context has confirmed to have done its job and its now `complete` / `used` and shouldn't get another attribution.

Reply:
I agree, something similar is implemented in d6b9e1b, although it doesn't set a `complete` / `used` flag on the filter but instead resets the Entity's context.